xc/xenctrlext.ml

```diff
@@ -83,5 +83,77 @@ external numainfo : handle -> numainfo = "stub_xenctrlext_numainfo"
 
 external cputopoinfo : handle -> cputopo array = "stub_xenctrlext_cputopoinfo"
 
-external xc_get_msr_arch_caps : handle -> int64
-  = "stub_xenctrlext_get_msr_arch_caps"
+module Policy = struct
+  open Xenctrl
+
+  let u32hex =
+    let from_hex s = Scanf.sscanf s "%Lx" @@ fun x -> x in
+    let to_hex = Printf.sprintf "%08Lx" in
+    Tyre.conv from_hex to_hex @@ Tyre.regex (Re.repn Re.xdigit 8 (Some 8))
+
+  let u64hex =
+    (* we cannot store the value as unsigned, but we should print it as such in hex *)
+    let from_hex s = Scanf.sscanf s "%Lx" @@ fun x -> x in
+    let to_hex = Printf.sprintf "%016Lx" in
+    Tyre.conv from_hex to_hex @@ Tyre.regex (Re.repn Re.xdigit 16 (Some 16))
+
+  let cpuid_leaf =
+    let leaf_of (((((leaf, subleaf), a), b), c), d) =
+      {leaf; subleaf; a; b; c; d}
+    in
+    let of_leaf l = (((((l.leaf, l.subleaf), l.a), l.b), l.c), l.d) in
+    let open Tyre in
+    let ( <:> ) a b = a <* str ":" <&> b in
+    let open Tyre.Infix in
+    Tyre.conv leaf_of of_leaf
+    @@ (u32hex
+       <:> u32hex
+       <* str "->"
+       <&> u32hex
+       <:> u32hex
+       <:> u32hex
+       <:> u32hex
+       )
+
+  let msr =
+    let msr_of ((idx, value), flags) = {idx; value; flags} in
+    let of_msr m = ((m.idx, m.value), m.flags) in
+    let open Tyre.Infix in
+    Tyre.conv msr_of of_msr
+    @@ (u32hex
+       <* Tyre.str "->"
+       <&> u64hex
+       <* Tyre.str "("
+       <&> u32hex
+       <* Tyre.str ")"
+       )
+
+  let comma_sep_array el =
+    let open Tyre in
+    conv Array.of_list Array.to_list @@ separated_list ~sep:(str ";") el
```

Comment:
comma, or semicolon separated?

Reply:
good point, but I dropped this code now.